sql/tests: TestRandomSyntaxSQLSmith failed [unexpectedly unordered keys in jsonObject]

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/14610717?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/14610717?buildTab=artifacts#/) on release-23.2.4-rc @ [c013b0d428ae23ac3542d950d9359cfc2a0246a5](https://github.com/cockroachdb/cockroach/commits/c013b0d428ae23ac3542d950d9359cfc2a0246a5):

Random syntax error:

```
    rsg_test.go:911: Crash detected: server panic: pq: internal error: unexpectedly unordered keys in jsonObject {"\u0005\u0014\")[\u001e": "_p\u0014", "3XN\u0008pTN": ""}
```
Query:

```
        WITH
        	"w😇i""th185000" (col1039069)
        		AS (
        			SELECT
        				*
        			FROM
        				(
        					VALUES
        						(ARRAY[]:::DECIMAL[]),
        						(ARRAY[3.826042721469873768:::DECIMAL,(-2495.891119824411118):::DECIMAL]),
        						(
        							ARRAY[(-2157482.35478645044):::DECIMAL,4323897123.139193817:::DECIMAL,9.193059313478565902E+22:::DECIMAL]
        						),
        						(
        							ARRAY[(-0.01958105118014557863):::DECIMAL,'-Infinity':::DECIMAL,(-744.4754516228360795):::DECIMAL]
        						)
        				)
        					AS tab454439 (col1039069)
        		),
        	w͊ith185001 (col1039070)
        		AS (
        			SELECT
        				*
        			FROM
        				(
        					VALUES
        						('{"b": 0.9599167020549824, "foo": true, "wmaJu": {"bar": false}}':::JSONB),
        						(
        							'[[[1.2002019188152038], false], null, [[null, false, false], false, true], true, null]':::JSONB
        						),
        						(
        							'[{"BNA5a": "c", "foo": false}, {"Fd4ujFoXd": 1.7420763588047599, "baz": "EO", "gyH2M": false}, true, false, "foo", 2.156513194049362]':::JSONB
        						),
        						(
        							json_object(ARRAY[e'3XN\bpTN':::STRING,'':::STRING,e'\x05\x14")[\x1e':::STRING,e'_p\x14':::STRING]::STRING[])::JSONB
        						)
        				)
        					AS tab454440 (col1039070)
        		)
        SELECT
        	jsonb_object_agg('[{"IeVFqMr": null, "XeyY": false, "a": 0.963213451556605, "wCTkU": [{"Kd": "c", "sTUkuoJuk": "srTVTn"}, "1zVgfh", null]}, {"2mpHmSU": 0.47216566718681685, "a": "i1P", "c": false, "foo": ["foobar"]}, null]':::JSONB::STRING, cte_re😐f53259.col1039070) OVER (PARTITION BY cte_re😐f53259.col1039070)::JSONB
        		AS col1039071
        FROM
        	w͊ith185001 AS cte_re😐f53259
        WHERE
        	true
        GROUP BY
        	cte_re😐f53259.col1039070
        LIMIT
        	21:::INT8;
```
Schema:

```
    rsg_test.go:720: To reproduce, use schema:
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;;
    rsg_test.go:722: CREATE TABLE table1 (col1_0 REFCURSOR NOT NULL, "col1%fe_\\U000C8DB91" TSQUERY, "col1 _2" INET NOT NULL, col1_3 REFCURSOR NULL, col1_4 DECIMAL NOT NULL, col1_5 BYTES NOT NULL, "Col1_6" GEOMETRY, INDEX (lower(CAST("col1 _2" AS STRING)) ASC) STORING (col1_0, "col1%fe_\\U000C8DB91", "Col1_6") WHERE (table1.col1_5 >= '\x00':::BYTES) OR (table1.col1_4 <= 'Infinity':::DECIMAL), INVERTED INDEX ("Col1_6") WHERE (table1.col1_4 >= (-1.234E+401):::DECIMAL) OR (table1.col1_5 < '\xe29883':::BYTES) NOT VISIBLE);
    rsg_test.go:722: CREATE TABLE "ta""ble2" ("co.l2_0" NAME NOT NULL, col2_̼1 TSQUERY NOT NULL, col2_2 BYTES NOT NULL, col2_3 TIMESTAMPTZ NOT NULL, col2_4 OID, "coL2_5" TIMETZ NOT NULL, PRIMARY KEY ("coL2_5" ASC), INDEX ("coL2_5", col2_4 DESC, col2_3 ASC) STORING (col2_̼1, col2_2) WHERE ((("ta""ble2"."co.l2_0" != e'\x00':::STRING:::NAME) OR ("ta""ble2".col2_2 = '\x':::BYTES)) AND ("ta""ble2".col2_3 < '4714-11-24 00:00:00+00 BC':::TIMESTAMPTZ)) AND ("ta""ble2"."coL2_5" < '00:00:00+15:59':::TIMETZ), INDEX ("coL2_5" ASC) STORING ("co.l2_0", col2_2) WHERE (("ta""ble2"."co.l2_0" < e'\'':::STRING:::NAME) OR ("ta""ble2".col2_2 < '\x22':::BYTES)) AND ("ta""ble2"."coL2_5" != '24:00:00-15:59':::TIMETZ), UNIQUE ("co.l2_0" ASC) STORING (col2_̼1, col2_4) WHERE "ta""ble2"."co.l2_0" != '"':::STRING:::NAME, INDEX (lower(CAST("coL2_5" AS STRING))), UNIQUE (col2_2 DESC, lower("co.l2_0"), col2_3 DESC) WHERE (("ta""ble2".col2_2 < '\x00':::BYTES) AND ("ta""ble2"."co.l2_0" = e'\'':::STRING:::NAME)) OR ("ta""ble2".col2_3 > '4714-11-24 00:00:00+00 BC':::TIMESTAMPTZ), UNIQUE ("co.l2_0", "coL2_5" ASC) STORING (col2_2) WHERE ((("ta""ble2".col2_2 != '\x27':::BYTES) AND ("ta""ble2"."co.l2_0" != e'\x00':::STRING:::NAME)) OR ("ta""ble2"."coL2_5" != '00:00:00+15:59':::TIMETZ)) OR ("ta""ble2".col2_3 <= '4714-11-24 00:00:00+00 BC':::TIMESTAMPTZ));
    rsg_test.go:722: ALTER TABLE table1 INJECT STATISTICS e'[{"avg_size": 13, "columns": ["col1_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1311400137244618782, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1914831732785225361}, {"avg_size": 11, "columns": ["\\"Col1_6\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 455652432457731525, "histo_buckets": [{"distinct_range": 0, "num_eq": 1000000000, "num_range": 0, "upper_bound": "\\\\x42fd052cd4000000000000c200001fa3ce9360c1f51c35c12729f342024e0c9fb53a2441e033f9484eeb50"}, {"distinct_range": 0, "num_eq": 800000, "num_range": 3035127757688952122, "upper_bound": "\\\\x42fd0cd000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 47856713.22601091, "num_eq": 60000000, "num_range": 60000000, "upper_bound": "\\\\x42fd0ce004000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 500, "num_eq": 4310052766534779015, "num_range": 500, "upper_bound": "\\\\x42fd0d4000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 0, "num_eq": 5000, "num_range": 0, "upper_bound": "\\\\x42fd0d9000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 6610520682817444000, "num_eq": 100000, "num_range": 6610520682817443997, "upper_bound": "\\\\x42fd100000000000000000c20204a3ce6f7af8c201e09272de4f8d42029444e264cfd8420105c7e71d308a"}, {"distinct_range": 546.4084800545268, "num_eq": 9000000, "num_range": 800, "upper_bound": "\\\\x42fd1000000000000040004024000000000000402400000000000040440000000000004044000000000000"}, {"distinct_range": 0, "num_eq": 80000000, "num_range": 4712606690475522195, "upper_bound": "\\\\x42fd10000000000000c0004024000000000000402400000000000040440000000000004044000000000000"}, {"distinct_range": 10, "num_eq": 100000, "num_range": 10, "upper_bound": "\\\\x42fd1000000000000110004024000000000000402400000000000040440000000000004044000000000000"}, {"distinct_range": 487.60128199791535, "num_eq": 4000, "num_range": 3000, "upper_bound": "\\\\x42fd10000000000001ab004024000000000000402400000000000040440000000000004044000000000000"}, {"distinct_range": 10000, "num_eq": 1000, "num_range": 10000, "upper_bound": "\\\\x42fd10000000000001d0004024000000000000402400000000000040440000000000004044000000000000"}, {"distinct_range": 1345263342232334600, "num_eq": 6697027367568957469, "num_range": 2145075316147848343, "upper_bound": "\\\\x42fd110000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 7897094294019802000, "num_eq": 50000000000, "num_range": 7897094294019801935, "upper_bound": "\\\\x42fd130000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 0, "num_eq": 0, "num_range": 1761194485154870397, "upper_bound": "\\\\x42fd170000000000000000c200001fa3ce9360c1f51c35c12729f342024e0c9fb53a2441e033f9484eeb50"}, {"distinct_range": 7487384117171413000, "num_eq": 50000000000, "num_range": 7487384117171413002, "upper_bound": "\\\\x42fd170000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 0, "num_eq": 4546126916794566815, "num_range": 10000, "upper_bound": "\\\\x42fd181000000000000000c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 0, "num_eq": 7250763538774944828, "num_range": 0, "upper_bound": "\\\\x42fd19c000000000000000c200001fa3ce9360c1f51c35c12729f342024e0c9fb53a2441e033f9484eeb50"}, {"distinct_range": 7963383404.753042, "num_eq": 90000000000, "num_range": 8000000000, "upper_bound": "\\\\x42fd1b0000000000000000c200001fa3ce9360c1f51c35c12729f342024e0c9fb53a2441e033f9484eeb50"}, {"distinct_range": 1000000, "num_eq": 8293517974135262361, "num_range": 1000000, "upper_bound": "\\\\x42fdffffffffffffffff00c1f9068309fed942c1fc9f8d512f9d1c4201bbf126db92bc4200f59810d008a6"}, {"distinct_range": 5046577853753851000, "num_eq": 40000000, "num_range": 5046577853753850611, "upper_bound": "\\\\x42fdffffffffffffffff00c200001fa3ce9360c1f51c35c12729f342024e0c9fb53a2441e033f9484eeb50"}, {"distinct_range": 6796909605155856000, "num_eq": 7380972955138046548, "num_range": 6796909605155855877, "upper_bound": "\\\\x42fdffffffffffffffff00c20192074496859ec1fef8c8982e99fa42016c9e5386aade42014c169a0a5d0c"}, {"distinct_range": 8.231026227401099, "num_eq": 10000000, "num_range": 300, "upper_bound": "\\\\x42fdffffffffffffffff00c20204a3ce6f7af8c201e09272de4f8d42029444e264cfd8420105c7e71d308a"}, {"distinct_range": 7000000, "num_eq": 5342307823170678113, "num_range": 7000000, "upper_bound": "\\\\x42fdffffffffffffffff01420168c430c35440c200b563c87fc9f0"}], "histo_col_type": "BYTES", "name": "__auto__", "null_count": 7867820573990664, "row_count": 1914831732785225361}, {"avg_size": 3, "columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 43023046512253346, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1914831732785225361}, {"avg_size": 28, "columns": ["\\"col1%fe_\\\\\\\\U000C8DB91\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1076669954449087547, "histo_col_type": "", "name": "__auto__", "null_count": 885041228515374324, "row_count": 1914831732785225361}, {"avg_size": 6, "columns": ["\\"col1 _2\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 678153176207290506, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1914831732785225361}, {"avg_size": 5, "columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1124706309026264355, "histo_col_type": "", "name": "__auto__", "null_count": 253429540779667259, "row_count": 1914831732785225361}, {"avg_size": 8, "columns": ["col1_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 30421428673498253, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1914831732785225361}]':::JSONB;
    rsg_test.go:722: ALTER TABLE "ta""ble2" INJECT STATISTICS e'[{"avg_size": 29, "columns": ["\\"coL2_5\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 194420781, "histo_buckets": [{"distinct_range": 0, "num_eq": 90000000000, "num_range": 0, "upper_bound": "00:00:00+15:59"}, {"distinct_range": 4468225757381388300, "num_eq": 7000000, "num_range": 4468225757381388085, "upper_bound": "05:00:05.218392+02:45"}, {"distinct_range": 90000000000, "num_eq": 90000000000, "num_range": 90000000000, "upper_bound": "11:27:46.765584+07:34"}, {"distinct_range": 0, "num_eq": 7755817622380689019, "num_range": 4398349010785936446, "upper_bound": "14:14:18.29159+10:03"}, {"distinct_range": 374499043068154300, "num_eq": 4022700463628657809, "num_range": 2757645391134506888, "upper_bound": "14:58:30.161855+04:54"}, {"distinct_range": 5142.133519988237, "num_eq": 5299814305627340044, "num_range": 6000, "upper_bound": "16:51:08.496606+03:16"}, {"distinct_range": 1640606868656025000, "num_eq": 1000000, "num_range": 1640606868656025051, "upper_bound": "21:12:35.714627+05:10"}], "histo_col_type": "TIMETZ", "name": "__auto__", "null_count": 0, "row_count": 200000000}, {"avg_size": 21, "columns": ["\\"co.l2_0\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 39895682, "histo_buckets": [{"distinct_range": 0, "num_eq": 7597205757904122222, "num_range": 0, "upper_bound": "\\rJ\\u0016@\\u0012Y"}, {"distinct_range": 9000000, "num_eq": 6818061979809504615, "num_range": 9000000, "upper_bound": "\\u0015(RQ"}, {"distinct_range": 0, "num_eq": 3000000, "num_range": 10000, "upper_bound": "-"}, {"distinct_range": 0, "num_eq": 900000000, "num_range": 0, "upper_bound": "Sb7\\u000ed"}, {"distinct_range": 2323536192.377636, "num_eq": 628416627774710070, "num_range": 3000000000, "upper_bound": "gD{m"}, {"distinct_range": 0, "num_eq": 10000000, "num_range": 4000, "upper_bound": "m\'y9\\u0005"}, {"distinct_range": 0, "num_eq": 60, "num_range": 4204508003688183249, "upper_bound": "|&F["}, {"distinct_range": 93150459.70206793, "num_eq": 8563227212655520078, "num_range": 300000000, "upper_bound": "\U00002603"}], "histo_col_type": "NAME", "name": "__auto__", "null_count": 0, "row_count": 200000000}, {"avg_size": 10, "columns": ["col2_\u033C1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 91228060, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 200000000}, {"avg_size": 7, "columns": ["col2_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 88332014, "histo_col_type": "BYTES", "name": "__auto__", "null_count": 0, "row_count": 200000000}, {"avg_size": 24, "columns": ["col2_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 188408588, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 200000000}, {"avg_size": 18, "columns": ["col2_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 52496885, "histo_col_type": "", "name": "__auto__", "null_count": 79572078, "row_count": 200000000}]':::JSONB;
    rsg_test.go:722: CREATE TYPE rand_typ_0 AS ENUM ('y', 'kw', 'ovpv');
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: 
        BEGIN; CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning'); COMMIT;
        BEGIN;
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        COMMIT;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (8840758865485822094);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (2006501665286497606);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (1412167556695033317);
    rsg_test.go:722: ALTER TABLE seed SCATTER;
    rsg_test.go:724: 
    rsg_test.go:725: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSQLSmith725101915
--- FAIL: TestRandomSyntaxSQLSmith (383.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #120970 sql/tests: TestRandomSyntaxSQLSmith failed [SIGSEGV in geos.LineMerge] [A-spatial C-bug C-test-failure O-robot P-3 T-sql-queries branch-release-23.1.18-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-37187

--- a/pkg/sql/logictest/testdata/logic_test/aggregate
+++ b/pkg/sql/logictest/testdata/logic_test/aggregate
@@ -3641,10 +3641,10 @@ SELECT json_object_agg(user_name, networks) FROM
 {"Alice": {"Facebook": "Alice_fb", "Instagram": "AliceInst", "Twitter": "@Alice"}}
 {"Bob": {"Facebook": "Bob_fb", "LinkedIn": "Bob The Builder"}}
 
-statement error pq: field name must not be null
+statement error pgcode 22004 null value not allowed for object key
 SELECT json_object_agg(null, null)
 
-statement error pq: field name must not be null
+statement error pgcode 22004 null value not allowed for object key
 SELECT json_object_agg(null, 1)
 
 statement ok

--- a/pkg/sql/logictest/testdata/logic_test/json_builtins
+++ b/pkg/sql/logictest/testdata/logic_test/json_builtins
@@ -1490,3 +1490,16 @@ bar   blah
 bar   blah2
 bar2  blah
 bar2  blah2
+
+# Regression test for incorrectly using jsonString.String() for asserting that
+# keys are ordered in the jsonObject (#121326).
+query T rowsort
+WITH cte(col) AS (
+  VALUES
+    ('false'::JSONB),
+    (jsonb_object(ARRAY['0', '', e'\x14', '']::TEXT[]))
+  )
+SELECT jsonb_object_agg('k', 'v') OVER (PARTITION BY cte.col) FROM cte;
+----
+{"k": "v"}
+{"k": "v"}

--- a/pkg/sql/logictest/testdata/logic_test/json_builtins
+++ b/pkg/sql/logictest/testdata/logic_test/json_builtins
@@ -577,20 +577,20 @@ SELECT json_build_object(''::void, 1)
 {"": 1}
 
 # even number of arguments
-query error pq: json_build_object\(\): argument list must have even number of elements
+statement error pgcode 22023 json_build_object\(\): argument list must have even number of elements
 SELECT json_build_object(1,2,3)
 
 # keys must be scalar and not null
-query error pq: json_build_object\(\): argument 1 cannot be null
+statement error pgcode 22004 json_build_object\(\): null value not allowed for object key
 SELECT json_build_object(null,2)
 
-query error pq: json_build_object\(\): key value must be scalar, not array, tuple, or json
+statement error pgcode 22023 json_build_object\(\): key value must be scalar, not array, composite, or json
 SELECT json_build_object((1,2),3)
 
-query error pq: json_build_object\(\): key value must be scalar, not array, tuple, or json
+statement error pgcode 22023 json_build_object\(\): key value must be scalar, not array, composite, or json
 SELECT json_build_object('{"a":1,"b":2}'::JSON, 3)
 
-query error pq: json_build_object\(\): key value must be scalar, not array, tuple, or json
+statement error pgcode 22023 json_build_object\(\): key value must be scalar, not array, composite, or json
 SELECT json_build_object('{1,2,3}'::int[], 3)
 
 query T
@@ -734,13 +734,13 @@ SELECT jsonb_build_array('+Inf'::FLOAT8, 'NaN'::FLOAT8)::STRING::JSONB
 ----
 ["Infinity", "NaN"]
 
-query error pq: json_object\(\): array must have even number of elements
+statement error pgcode 2202E json_object\(\): array must have even number of elements
 SELECT json_object('{a,b,c}'::TEXT[])
 
-query error pq: json_object\(\): null value not allowed for object key
+statement error pgcode 22004 json_object\(\): null value not allowed for object key
 SELECT json_object('{NULL, a}'::TEXT[])
 
-query error pq: json_object\(\): null value not allowed for object key
+statement error pgcode 22004 json_object\(\): null value not allowed for object key
 SELECT json_object('{a,b,NULL,"d e f"}'::TEXT[],'{1,2,3,"a b c"}'::TEXT[])
 
 query error pq: json_object\(\): mismatched array dimensions
@@ -787,13 +787,13 @@ SELECT json_object('{a,b,c,"d e f"}'::TEXT[],'{1,2,3,"a b c"}'::TEXT[])
 ----
 {"a": "1", "b": "2", "c": "3", "d e f": "a b c"}
 
-query error pq: jsonb_object\(\): array must have even number of elements
+statement error pgcode 2202E jsonb_object\(\): array must have even number of elements
 SELECT jsonb_object('{a,b,c}'::TEXT[])
 
-query error pq: jsonb_object\(\): null value not allowed for object key
+statement error pgcode 22004 jsonb_object\(\): null value not allowed for object key
 SELECT jsonb_object('{NULL, a}'::TEXT[])
 
-query error pq: jsonb_object\(\): null value not allowed for object key
+statement error pgcode 22004 jsonb_object\(\): null value not allowed for object key
 SELECT jsonb_object('{a,b,NULL,"d e f"}'::TEXT[],'{1,2,3,"a b c"}'::TEXT[])
 
 query error pq: jsonb_object\(\): mismatched array dimensions

--- a/pkg/sql/sem/builtins/aggregate_builtins.go
+++ b/pkg/sql/sem/builtins/aggregate_builtins.go
@@ -5356,8 +5356,8 @@ func (a *jsonObjectAggregate) Add(
 
 	// If the key datum is NULL, return an error.
 	if datum == tree.DNull {
-		return pgerror.New(pgcode.InvalidParameterValue,
-			"field name must not be null")
+		return pgerror.New(pgcode.NullValueNotAllowed,
+			"null value not allowed for object key")
 	}
 
 	key, err := asJSONBuildObjectKey(

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -9314,9 +9314,9 @@ var (
 )
 
 var (
-	errJSONObjectNotEvenNumberOfElements = pgerror.New(pgcode.InvalidParameterValue,
+	errJSONObjectNotEvenNumberOfElements = pgerror.New(pgcode.ArraySubscript,
 		"array must have even number of elements")
-	errJSONObjectNullValueForKey = pgerror.New(pgcode.InvalidParameterValue,
+	errJSONObjectNullValueForKey = pgerror.New(pgcode.NullValueNotAllowed,
 		"null value not allowed for object key")
 	errJSONObjectMismatchedArrayDim = pgerror.New(pgcode.InvalidParameterValue,
 		"mismatched array dimensions")
@@ -9548,8 +9548,7 @@ var jsonBuildObjectImpl = tree.Overload{
 		builder := json.NewObjectBuilder(len(args) / 2)
 		for i := 0; i < len(args); i += 2 {
 			if args[i] == tree.DNull {
-				return nil, pgerror.Newf(pgcode.InvalidParameterValue,
-					"argument %d cannot be null", i+1)
+				return nil, errJSONObjectNullValueForKey
 			}
 
 			key, err := asJSONBuildObjectKey(
@@ -11136,7 +11135,7 @@ func asJSONBuildObjectKey(
 	switch t := d.(type) {
 	case *tree.DArray, *tree.DJSON, *tree.DTuple:
 		return "", pgerror.New(pgcode.InvalidParameterValue,
-			"key value must be scalar, not array, tuple, or json")
+			"key value must be scalar, not array, composite, or json")
 	case *tree.DCollatedString:
 		return t.Contents, nil
 	case *tree.DString:

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -2078,6 +2078,8 @@ func (expr *FuncExpr) MaybeWrapError(err error) error {
 		return err
 	}
 	// Otherwise, wrap it with context.
+	// TODO(yuzefovich): consider removing this context in order to match
+	// postgres error messages.
 	newErr := errors.Wrapf(err, "%s()", errors.Safe(fName))
 	// Count function errors as it flows out of the system. We need to handle
 	// them this way because if we are facing a retry error, in particular those

--- a/pkg/util/json/encoded.go
+++ b/pkg/util/json/encoded.go
@@ -594,17 +594,6 @@ func (j *jsonEncoded) AsArray() ([]JSON, bool) {
 	return decoded.AsArray()
 }
 
-func (j *jsonEncoded) AreKeysSorted() bool {
-	if dec := j.alreadyDecoded(); dec != nil {
-		return dec.AreKeysSorted()
-	}
-	decoded, err := j.decode()
-	if err != nil {
-		return false
-	}
-	return decoded.AreKeysSorted()
-}
-
 func (j *jsonEncoded) Compare(other JSON) (_ int, err error) {
 	if other == nil {
 		return -1, nil

--- a/pkg/util/json/json.go
+++ b/pkg/util/json/json.go
@@ -953,7 +953,7 @@ func (j jsonArray) AreKeysSorted() bool {
 func (j jsonObject) AreKeysSorted() bool {
 	keys := make([]string, 0, j.Len())
 	for _, a := range j {
-		keys = append(keys, a.k.String())
+		keys = append(keys, string(a.k))
 	}
 	return sort.StringsAreSorted(keys)
 }

--- a/pkg/util/json/json.go
+++ b/pkg/util/json/json.go
@@ -190,11 +190,6 @@ type JSON interface {
 	// and a boolean indicating if this JSON value is a array type.
 	AsArray() ([]JSON, bool)
 
-	// AreKeysSorted returns if the keys in a JSON Object are sorted by
-	// increasing order. It returns false if the underlying JSON value
-	// is not a JSON object.
-	AreKeysSorted() bool
-
 	// Exists implements the `?` operator: does the string exist as a top-level
 	// key within the JSON value?
 	//
@@ -924,38 +919,6 @@ func (j jsonArray) AsArray() ([]JSON, bool) {
 
 func (j jsonNumber) AsArray() ([]JSON, bool) {
 	return nil, false
-}
-
-func (j jsonNull) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonString) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonFalse) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonTrue) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonNumber) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonArray) AreKeysSorted() bool {
-	return false
-}
-
-func (j jsonObject) AreKeysSorted() bool {
-	keys := make([]string, 0, j.Len())
-	for _, a := range j {
-		keys = append(keys, string(a.k))
-	}
-	return sort.StringsAreSorted(keys)
 }
 
 // parseJSONGoStd parses json using encoding/json library.
@@ -2023,7 +1986,11 @@ func (j jsonObject) EncodeForwardIndex(buf []byte, dir encoding.Direction) ([]by
 	buf = encoding.EncodeJSONValueLength(buf, dir, int64(len(j)))
 
 	if buildutil.CrdbTestBuild {
-		if ordered := j.AreKeysSorted(); !ordered {
+		keys := make([]string, 0, j.Len())
+		for _, a := range j {
+			keys = append(keys, string(a.k))
+		}
+		if !sort.StringsAreSorted(keys) {
 			return nil, errors.AssertionFailedf("unexpectedly unordered keys in jsonObject %s", j)
 		}
 	}

